### PR TITLE
Fix performance regression caused by enabling opal thread support

### PR DESCRIPTION
**Pull request by hjelmn**
This commit adds opal_using_threads() protection around the atomic
operation in OBJ_RETAIN/OBJ_RELEASE. This resolves the performance
issues seen when running psm with MPI_THREAD_SINGLE.

To avoid issues with header dependencies opal_using_threads() has been
moved to a new header (thread_usage.h). The OPAL_THREAD_ADD\* and
OPAL_THREAD_CMPSET\* macros have also been relocated to this header.

This was cherry-picked off a fix applied to v1.8 and not master. See
open-mpi/ompi#1902.

(cherry picked from commit ce91307918eadcdb6e4a3c8b6429b36d206cc0e3)

Signed-off-by: Nathan Hjelm hjelmn@lanl.gov

### CODE REVIEW

**Comment by ibm-ompi** [comment]
Build Failed with XL compiler! Please review the log, and get in touch if you have questions.

 Gist: https://gist.github.com/ibm-ompi/7f110ab6c94da788f0fed01d83c12215

**Comment by ibm-ompi** [comment]
Build Failed with GNU compiler! Please review the log, and get in touch if you have questions.

 Gist: https://gist.github.com/ibm-ompi/514bb6fe123a40de30f56e64a600d85b

**Comment by mellanox-github** [comment]
Test PASSed.
See http://bgate.mellanox.com/jenkins/job/gh-ompi-release-pr/1989/ for details.

**Comment by hjelmn** [comment]
:bot:milestone:v2.0.1
:bot:label:bug
:bot:assign: @nysal 

This fix was lost last year before the 2.x branch.

**Comment by hjelmn** [comment]
@nysal. Please review.

**Comment by hjelmn** [comment]
:bot:milestone:v2.0.1
:bot:label:bug

**Comment by hjelmn** [comment]
:bot:assign: @nysal

File: opal/threads/thread_usage.h [hunk 1]
```diff
@@ -0,0 +1,213 @@
...
+        ret = opal_atomic_add_32(addr, delta);
+    } else {
+        ret = (*addr += delta);
+    }
+
+    return ret;
+}
+#else
+static inline int32_t
+OPAL_THREAD_ADD32(volatile int32_t *addr, int delta)
```
Comment (nysal) [inline opal/threads/thread_usage.h]: @hjelmn do we need two versions of OPAL_THREAD_ADD32, OPAL_THREAD_ADD64 and OPAL_THREAD_ADD_SIZE_T. Wouldn't the compiler remove the dead code if opal_using_threads() returns false?

**Comment by hjelmn** [comment]
I added the cleanup commit to this PR due to a bug in OPAL_THREAD_ADD64. The argument was of type int when it should have been int64_t.

**Comment by lanl-ompi** [comment]
Test FAILed.

**Comment by lanl-ompi** [comment]
Test FAILed.

**Comment by ibm-ompi** [comment]
Build Failed with GNU compiler! Please review the log, and get in touch if you have questions.

 Gist: https://gist.github.com/431e65fab860a6cb7d4a3438cc46aaeb

**Comment by mellanox-github** [comment]
Test FAILed.
See http://bgate.mellanox.com/jenkins/job/gh-ompi-release-pr/1995/ for details.

**Comment by lanl-ompi** [comment]
Test FAILed.

**Comment by ibm-ompi** [comment]
Build Failed with XL compiler! Please review the log, and get in touch if you have questions.

 Gist: https://gist.github.com/cbddb0dcc3fb86f444f37d60861b46c4

**Comment by nysal** [comment]
The XL compiler build failed with:

```
In file included from dss/dss_copy.c:24:
In file included from ../opal/util/output.h:70:
In file included from ../opal/class/opal_object.h:128:
../opal/threads/thread_usage.h:144:1: warning: implicit declaration of function 'OPAL_UNLIKELY' is invalid in C99 [-Wimplicit-function-declaration]
OPAL_THREAD_DEFINE_ATOMIC_ADD(int32_t, 32)
^
../opal/threads/thread_usage.h:107:9: note: expanded from macro 'OPAL_THREAD_DEFINE_ATOMIC_ADD'
    if (OPAL_UNLIKELY(opal_using_threads())) {                          \
        ^
1 warning generated.
1 warning generated.
```

The opal/prefetch.h header is probably not included

**Comment by hjelmn** [comment]
@nysal Yeah. Fixed that. I brought over one more commit because OPAL_THREAD_ADD64 and OPAL_THREAD_ADD_SIZE_T have the wrong argument type for delta. Been wrong for years. Figured we should go ahead and fix it now.

**Comment by mellanox-github** [comment]
Test PASSed.
See http://bgate.mellanox.com/jenkins/job/gh-ompi-release-pr/1996/ for details.

File: opal/threads/thread_usage.h [hunk 2]
```diff
@@ -0,0 +1,186 @@
...
+ * 
+ * $HEADER$
+ */
+
+#if !defined(OPAL_THREAD_USAGE_H)
+#define OPAL_THREAD_USAGE_H
+
+#include "opal/sys/atomic.h"
+#include "opal/prefetch.h"
+#include "opal_config.h"
```
Comment (nysal) [inline opal/threads/thread_usage.h:30]: Should we move "#include "opal_config.h"" to the top, although "opal/sys/atomic.h" does include it.

Comment (jsquyres) [inline opal/threads/thread_usage.h:30]: opal_config.h should always be the first header file included in any file.

Comment (hjelmn) [inline opal/threads/thread_usage.h:30]: Figured it was pointless to include since it is included through atomic.h. Will update.

Comment (jsquyres) [inline opal/threads/thread_usage.h:30]: Should you add a `#if !defined(...`) kind of check at the top to guarantee that this file is _only_ included via `atomic.h`?

**Comment by nysal** [comment]
With the previous comment about "opal_config.h" addressed, it looks good to me :+1:

**Comment by jsquyres** [comment]
@hppritcha Good to go

**Pull request merged**